File: native/Sources/XRDApp/ReportBuilder.swift
import Foundation
import XRDCore

/// Builds the per-file key/value report CSV. Shared by the single-file export
/// (DetailView) and batch export (AppModel) so they stay identical.
enum ReportBuilder {
    static func csv(displayName: String, fileName: String, result r: DGResult,
                    span: DGRange?, quality: ImpurityScan?,
                    crystallinity c: CrystallinityResult? = nil) -> String {
        var rows: [(String, String)] = [
            ("sample", displayName), ("file", fileName),
            ("method", r.methodName), ("wavelength_angstrom", String(format: "%.5f", r.wavelength)),
            ("DG_percent", String(format: "%.2f", r.dgPercent)),
            ("DG_sigma", r.dgSigma.map { String(format: "%.2f", $0) } ?? ""),
        ]
        if let s = span { rows += [("DG_range_low", String(format: "%.2f", s.low)),
                                   ("DG_range_high", String(format: "%.2f", s.high))] }
        rows += [
            ("peak_count", "\(r.peakCount)"),
            ("graphitic_2theta", String(format: "%.4f", r.graphitic.xc)),
            ("graphitic_FWHM", String(format: "%.4f", r.graphitic.w)),
            ("graphitic_mu", String(format: "%.4f", r.graphitic.mu)),
            ("graphitic_d_nm", String(format: "%.5f", r.graphitic.dSpacing)),
        ]
        if let t = r.turbostratic {
            rows += [("turbostratic_2theta", String(format: "%.4f", t.xc)),
                     ("turbostratic_FWHM", String(format: "%.4f", t.w)),
                     ("turbostratic_d_nm", String(format: "%.5f", t.dSpacing))]
        }
        rows += [
            ("X_graphitic", String(format: "%.4f", r.areaFractionGraphitic)),
            ("X_turbostratic", String(format: "%.4f", r.areaFractionTurbostratic)),
            ("d_prime_nm", String(format: "%.5f", r.dPrimeWeighted)),
            ("crystallite_Lc_A", String(format: "%.1f", r.crystalliteLc)),
            ("baseline_y0", String(format: "%.3f", r.y0)),
            ("two_theta_offset", String(format: "%+.3f", r.twoThetaOffset)),
            ("fit_R2", String(format: "%.5f", r.fitR2)),
        ]
        if let c = c {
            rows += [("crystalline_fraction_pct", String(format: "%.1f", c.crystallineFraction * 100)),
                     ("disordered_fraction_pct", String(format: "%.1f", c.disorderedFraction * 100)),
                     ("crystallinity_fit_R2", String(format: "%.4f", c.fitR2))]
        }
        if let q = quality { rows.append(("data_quality", q.verdict)) }
        return "key,value\n" + rows.map { "\($0.0),\"\($0.1)\"" }.joined(separator: "\n") + "\n"
    }

    /// Columns for the consolidated runs CSV (mirrors the Compare tab).
    static func consolidatedHeader() -> String {
        "file,carbon_type,carbon_ratio,fe_ratio,caco3_ratio,temperature_C,time_h,form,wash,DG,crystallinity_pct,disordered_pct,Lc,d_prime,graphitic_xc\n"
    }
    static func consolidatedRow(fileName: String, info: RunInfo?, result r: DGResult?,
                                crystallinity c: CrystallinityResult? = nil) -> String {
        func n(_ v: Double?) -> String { v.map { String(format: "%g", $0) } ?? "" }
        func q(_ s: String) -> String { s.contains(",") ? "\"\(s)\"" : s }
        let i = info
        var cols: [String] = [q(fileName), i?.carbonType ?? "", n(i?.carbonRatio), n(i?.feRatio),
                              n(i?.caco3Ratio), i?.temperatureC.map(String.init) ?? "", n(i?.timeH),
                              i?.form ?? "", i?.wash ?? ""]
        cols.append(r.map { String(format: "%.2f", $0.dgPercent) } ?? "")
        cols.append(c.map { String(format: "%.1f", $0.crystallineFraction * 100) } ?? "")
        cols.append(c.map { String(format: "%.1f", $0.disorderedFraction * 100) } ?? "")
        cols.append(r.map { String(format: "%.1f", $0.crystalliteLc) } ?? "")
        cols.append(r.map { String(format: "%.5f", $0.dPrimeWeighted) } ?? "")
        cols.append(r.map { String(format: "%.4f", $0.graphitic.xc) } ?? "")
        return cols.joined(separator: ",") + "\n"
    }
}

File: windows/XRDAnalyzer/Services/ReportBuilder.cs
using System.Globalization;
using System.Text;
using XRDAnalyzer.Engine.Models;

namespace XRDAnalyzer.Services;

/// Builds the per-file key/value report CSV and the consolidated runs CSV.
/// Mirrors `ReportBuilder` (native/Sources/XRDApp/ReportBuilder.swift) exactly
/// so exports from either platform have the same shape.
public static class ReportBuilder
{
    private static string F(double v, int decimals) => v.ToString("F" + decimals, CultureInfo.InvariantCulture);
    private static string Signed(double v) => (v >= 0 ? "+" : "") + v.ToString("F3", CultureInfo.InvariantCulture);

    public static string Csv(string displayName, string fileName, DGResultDto r,
        DGRangeDto? span, ImpurityScanDto? quality, CrystallinityDto? crystallinity = null)
    {
        var rows = new List<(string Key, string Value)>
        {
            ("sample", displayName), ("file", fileName),
            ("method", r.MethodName), ("wavelength_angstrom", F(r.Wavelength, 5)),
            ("DG_percent", F(r.DgPercent, 2)),
            ("DG_sigma", r.DgSigma is { } sg ? F(sg, 2) : ""),
        };
        if (span is not null)
        {
            rows.Add(("DG_range_low", F(span.Low, 2)));
            rows.Add(("DG_range_high", F(span.High, 2)));
        }
        rows.Add(("peak_count", r.PeakCount.ToString()));
        rows.Add(("graphitic_2theta", F(r.Graphitic.Xc, 4)));
        rows.Add(("graphitic_FWHM", F(r.Graphitic.W, 4)));
        rows.Add(("graphitic_mu", F(r.Graphitic.Mu, 4)));
        rows.Add(("graphitic_d_nm", F(r.Graphitic.DSpacing, 5)));
        if (r.Turbostratic is { } t)
        {
            rows.Add(("turbostratic_2theta", F(t.Xc, 4)));
            rows.Add(("turbostratic_FWHM", F(t.W, 4)));
            rows.Add(("turbostratic_d_nm", F(t.DSpacing, 5)));
        }
        rows.Add(("X_graphitic", F(r.AreaFractionGraphitic, 4)));
        rows.Add(("X_turbostratic", F(r.AreaFractionTurbostratic, 4)));
        rows.Add(("d_prime_nm", F(r.DPrimeWeighted, 5)));
        rows.Add(("crystallite_Lc_A", F(r.CrystalliteLc, 1)));
        rows.Add(("baseline_y0", F(r.Y0, 3)));
        rows.Add(("two_theta_offset", Signed(r.TwoThetaOffset)));
        rows.Add(("fit_R2", F(r.FitR2, 5)));
        if (crystallinity is { } c)
        {
            rows.Add(("crystalline_fraction_pct", F(c.CrystallineFraction * 100, 1)));
            rows.Add(("disordered_fraction_pct", F(c.DisorderedFraction * 100, 1)));
            rows.Add(("crystallinity_fit_R2", F(c.FitR2, 4)));
        }
        if (quality is not null) rows.Add(("data_quality", quality.Verdict));

        var sb = new StringBuilder("key,value\n");
        foreach (var (key, value) in rows) sb.Append(key).Append(",\"").Append(value).Append("\"\n");
        return sb.ToString();
    }

    public static string ConsolidatedHeader() =>
        "file,carbon_type,carbon_ratio,fe_ratio,caco3_ratio,temperature_C,time_h,form,wash,DG,crystallinity_pct,disordered_pct,Lc,d_prime,graphitic_xc\n";

    public static string ConsolidatedRow(string fileName, RunInfoDto? info, DGResultDto? r, CrystallinityDto? crystallinity = null)
    {
        static string N(double? v) => v?.ToString("G", CultureInfo.InvariantCulture) ?? "";
        static string Q(string s) => s.Contains(',') ? $"\"{s}\"" : s;

        var cols = new List<string>
        {
            Q(fileName), info?.CarbonType ?? "", N(info?.CarbonRatio), N(info?.FeRatio), N(info?.Caco3Ratio),
            info?.TemperatureC?.ToString() ?? "", N(info?.TimeH), info?.Form ?? "", info?.Wash ?? "",
            r is not null ? F(r.DgPercent, 2) : "",
            crystallinity is not null ? F(crystallinity.CrystallineFraction * 100, 1) : "",
            crystallinity is not null ? F(crystallinity.DisorderedFraction * 100, 1) : "",
            r is not null ? F(r.CrystalliteLc, 1) : "",
            r is not null ? F(r.DPrimeWeighted, 5) : "",
            r is not null ? F(r.Graphitic.Xc, 4) : "",
        };
        return string.Join(",", cols) + "\n";
    }
}
